metadata/ota_metadata/metadata_gen.py

```diff
@@ -191,15 +191,57 @@ def gen_metadata(
     target_abs = Path(os.path.abspath(target_dir))
     ignore = ignore_rules(target_dir, ignore_file)
 
+    # If ignore file has the following directories:
+    #    1, "home/autoware/*/build"
+    #    2, "home/autoware/*/src"
+    # We will NOT simply ignore files under them.
+    # We will check the following:
+    #    1, If the file is set as the target of a symblink ?
+    #    2, If the file falls in some special folder pattern ?
+    # Why we need to do this? It is explained in this document
+    # https://tier4.atlassian.net/wiki/x/JoC21Q
+    check_patterns = [r"home/autoware/[^/]+/build", r"home/autoware/[^/]+/src"]
+
+    # This is the flag to control if we will check and add files back to "build" and "src" folder
+    check_symlink = any(
+        bool(re.search(pattern, str(rule)))
+        for rule in ignore.rules
+        for pattern in check_patterns
+    )
+
+    # Special Patterns that we need to check and add files.
+    build_folder_patterns = [
+        r"home/autoware/[^/]*/build/.*/hook/.*",
+        r"home/autoware/[^/]*/build/.*/.*.egg-info/.*",
+    ]
```

Comment:
The same as above, we can pre-compile the patterns to speed up regex operations.